### PR TITLE
doc: documentation for ALTER SINK

### PR DESCRIPTION
**Pull request by petrosagg**
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: doc/user/content/sql/alter-sink.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+leads to a valid sink definition with the same rules as the `CREATE SINK`
+statement.
+
+When using the Avro format with a schema registry care must be taken such that
+the generated Avro schema of the new item is compatible with the previously
+published schema. When that is not the case the `ALTER SINK` statement will
+succeed but the subsequent execution of the sink will result in errors and will
+be unable to make progress.
+
+### Possible anomalies
```
Comment (sjwiesman) [inline doc/user/content/sql/alter-sink.md:42]: These are all standard operating procedure for a stream processor. I wonder if we want to call that out. We're "only" providing the same level of correctness as a sp here as opposed to our standard higher level of correctness.

File: doc/user/content/sql/alter-sink.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+**Anomaly 1: Topic contains stale value for a key**
+
+As a consequence of not re-emitting a snapshot after `ALTER SINK`, all keys of
+the collection will appear to have the value they had in the old collection
+until an update happens to them. At that point the current value will be
+published to the topic.
+
+It is important that readers of the topic are prepared to handle an old value
+for a key, for example by filling in additional columns with default values.
+Alternatively, forcing an update to all the keys after `ALTER SINK` will force
```
Comment (sthm) [inline doc/user/content/sql/alter-sink.md:58]: Does it make sense to mention that dropping and recreating the sink emits a complete snapshot to guild folks who should NOT use `ALTER SINK in` the right direction?

Comment (morsapaes) [reply doc/user/content/sql/alter-sink.md:58]: Is there a use case for `ALTER SINK` outside blue/green deployments right now? I'd like to frame this in the clearest possible way. If there isn't, no one should be manually running this command, and these docs should point very early on to the dbt workflow. I'm pushing some updates soon.

**Comment by morsapaes** [comment]
Pushed a commit with a first review pass. I think we shouldn't refer to the different possible scenarios as "anomalies", but rather "cutover scenarios"; it'd be great to provide some more concrete examples of what the data would look like for each of these scenarios, since otherwise it's a tad bit hard to understand.

**Review by morsapaes — APPROVED** [review]
Also, approving since this looks like a good base and we can iterate on the documentation as users start picking this up. I still need to write the dbt workflow counterpart, but will push it into #27557 or as a separate PR.

File: doc/user/content/sql/alter-sink.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+causing disruption to downstream consumers. This is useful in the context
+of [blue/green deployments](/manage/dbt/development-workflows/#bluegreen-deployments).
+
+## Syntax
+
+{{< diagram "alter-sink.svg" >}}
+
+## Details
+
+To alter the upstream relation a sink depends on and ensure continuity in data
```
Comment (benesch) [inline doc/user/content/sql/alter-sink.md:19]:
```suggestion
To alter a sink's upstream relation and ensure continuity in data
```

(I think "upstream" and "depends on" are redundant here.)

Comment (benesch) [reply doc/user/content/sql/alter-sink.md:19]: (Also @morsapaes the `CREATE SINK` documentation says "underlying relation" where we say "upstream relation" here—we should bring them in sync! I think I like the new "upstream relation" term better.)

Comment (benesch) [reply doc/user/content/sql/alter-sink.md:19]: Actually I feel like something about the wording is off here! It's not about altering _and_ ensuring continuity, it's about ensuring continuity _while_ altering, right?

```suggestion
To ensure continuity in data processing when altering a sink's upstream relation, Materialize must pick ...
```

File: doc/user/content/sql/alter-sink.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+processing, Materialize must pick a consistent cutover timestamp. When you
+execute an `ALTER SINK` command, the resulting output will contain all the
+updates that happened before the cutover timestamp for the old relation, as
+well as all the updates that happened after the cutover timestamp for the new
+relation.
+
+{{< note >}}
+To select a consistent timestamp, Materialize waits for the previous definition
+of the sink to make enough progress such that there is overlap between the
+available timestamps of the newly specified upstream relation. Attempting to
```
Comment (benesch) [inline doc/user/content/sql/alter-sink.md:29]: The wording of the first sentence here is a little wonky. Tinkering a bit:

```suggestion
To select a consistent timestamp, Materialize must wait for the previous definition
of the sink to emit results up until the oldest timestamp at which the contents of
the new upstream relation are known. Attempting to
```

File: doc/user/content/sql/alter-sink.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+For `ALTER SINK` to be successful, the newly specified relation must lead to a
+valid sink definition with the same conditions as the original `CREATE SINK`
+statement.
+
+When using the Avro format with a schema registry, the generated Avro
+schema for the new relation must be compatible with the previously published
+schema. If that's not the case, the `ALTER SINK` command will succeed, but the
+subsequent execution of the sink will result in errors and will not be able to
+make progress.
```
Comment (benesch) [inline doc/user/content/sql/alter-sink.md:44]: Maybe add a recommendation here: monitor your sink status via <this SQL query> or in the console after an alteration to be ensure it remains healthy.

**Review by benesch — APPROVED** [review]
Very nice.

**Pull request merged**